### DISCUSSION TITLE
What’s the recommended way to get selected text in the editor

### DISCUSSION THREAD

**Discussion by Oliver-Weng-Sorcerer** [coder/code-server#5077] [Q&A] [answered]
Currently I’m working on an additional feature, wondering what’s the best way to get the selected text in the editor.

Is there any Js object exposed in root document something like “editor.getSlection()” exists?

currently I’m using document.selection().toString(), but that doesn’t work well with Firefox.

Any ideas ?

**Comment by Oliver-Weng-Sorcerer** [comment]
@code-asher

**Comment by code-asher** [comment, marked as answer]
I am not sure JavaScript-level features will work very well here.  The
best method is to go through the Code API via a custom extension.

https://code.visualstudio.com/api/references/vscode-api

For example (untested):

```
const editor = vscode.window.activeTextEditor
conts text = editor.getText(editor.selection)
````

**Reply by Oliver-Weng-Sorcerer** [reply]
it looks like the vscode is not exposed to the window object, how do I access this vscode object ? assuming it needs to be in a particular runtime ?

**Reply by code-asher** [reply]
Yeah the Code API can only be accessed through an extension.

https://code.visualstudio.com/api/get-started/your-first-extension

**Reply by Oliver-Weng-Sorcerer** [reply]
awesome, thanks for the tips.

**Reply by Oliver-Weng-Sorcerer** [reply]
HI @code-asher ,

I have built a vscode extension and published to the open vsx registry.

but when I try to use the command 

code-server --install-extension xxxxx
it says: `Incompatible: Can't install xxxxxxxx.xxxxxxx extension because it is not compatible with the current version of code-server (version 1.66.2).`

any insight about this one ?

**Reply by Oliver-Weng-Sorcerer** [reply]
I got the version issue resolved, now the log says the extension is successfully installed but it doesn't show as in installed extension panel. 
am I missing any steps ?

**Reply by jsjoeio** [reply]
Hmm...do you have any logs in the console? It should show up. Happy to test as well if the extension is open source.

**Reply by Oliver-Weng-Sorcerer** [reply]
@jsjoeio 
Hi there, it's all set, the issue was: the code server was installed under a particular user, but then the extension was installed under root user. as soon as I installed the extension under the user level, it works.

**Reply by jsjoeio** [reply]
ohhh! hooray!

**Comment by Oliver-Weng-Sorcerer** [comment]
Hi @jsjoeio and @code-asher ,

I'm not sure if it's just me or it's a bug.

here is what I see right now.

my extension has this code snippet in it. `vscode.env.clipboard.readText();`  it reads the system clipboard from vscode object. 

but whenever I use the extension in the code server in Firefox, and start writing content in a file, it gives me this error. but it's happening in chrome.

I'm wondering if code server internally interpret the vscode.env.clipboard object to navigator.clipboard, since it's currently a firefox limitation. 

Any workaround to avoid the warning message downbelow?

ref: https://developer.mozilla.org/en-US/docs/Web/API/Clipboard/read 

<img width="529" alt="Screen Shot 2022-06-22 at 4 44 44 PM" src="https://user-images.githubusercontent.com/28814939/175132748-a59d878e-f640-46d5-ad1d-b7b6a0a1cc52.png">

**Reply by jsjoeio** [reply]
🤔 Hmm...usually there is a popup asking if you want to give clipboard access to a website. I believe it's based on each URL. Here's an example in Brave's Settings.

<img width="1052" alt="image" src="https://user-images.githubusercontent.com/3806031/175134191-c1b1b4ac-3a49-4b68-8b90-99b6934891df.png">

If you're seeing that popup, then it sounds like the site you're browsing doesn't have access. Have you checked at the browser-level if this is the case?

**Reply by Oliver-Weng-Sorcerer** [reply]
@jsjoeio , I don't see any pop up when using it in firefox.

it doesn't give me any window to allow it.

wondering if you have time to verify it on your end with firefox ? to see if an extension with code `vscode.env.clipboard.readText();`  on `vscode.workspace.onDidChangeTextDocument` , and see if the firefox gives any popup?

**Reply by jsjoeio** [reply]
If you have a `vsix` file you can send me or attach here, I can try it!

**Reply by Oliver-Weng-Sorcerer** [reply]
Hi @jsjoeio , I can't attach a vsix file via github. So I sent it via email attachment.

**Reply by jsjoeio** [reply]
Got it! Installed the extension. Now how do I reproduce your issue?

<img width="1904" alt="image" src="https://user-images.githubusercontent.com/3806031/175577821-534b9667-7631-4f85-9a38-f69eadbc17f8.png">

**Reply by Oliver-Weng-Sorcerer** [reply]
you just need to use latest firefox and then open a file in code editor, and start typing, and see if you get those annoying pop ups

**Reply by jsjoeio** [reply]
<img width="1712" alt="image" src="https://user-images.githubusercontent.com/3806031/176024364-8baba1c0-c3d9-4574-a3ec-db4250dee559.png">

yes i do!

**Reply by Oliver-Weng-Sorcerer** [reply]
yea, that's odd, is this a bug? coz theoretically if I use `vscode.env.clipboard.readText();` in extension code, the code server should be able to handle it just like how it does in chrome.

**Reply by jsjoeio** [reply]
can you reproduce it with Codespaces or vscode.dev? i am guessing you need to request permissions or something before you can call `readText`. I believe that's how it works in the browser. 

I found this in the VSCode source code: https://github.com/microsoft/vscode/blob/ffe53e8d710a6fce5e2bdc67f25914aea3c40d42/src/vs/editor/contrib/clipboard/browser/clipboard.ts#L26-L28

Is your extension marked as a browser/web extension?

**Reply by Oliver-Weng-Sorcerer** [reply]
how do I mark my extension as browser/web extension ?

**Reply by jsjoeio** [reply]
See: https://code.visualstudio.com/api/extension-guides/web-extensions#update-existing-extensions-to-web-extensions

**Reply by Oliver-Weng-Sorcerer** [reply]
Hi @jsjoeio 

I ran some experiment with the web extension option, that's not going to work for me, because I'm using some of the node libraries  which is only available in regular extension runtime.

but what's interesting is that the extension works well in chrome with code server, but not working well in firefox. (I tried vscode.dev it doesn't allow me to install the extension, it says "extension is not available in Visual Studio Code for the Web." )

I'm wondering what's the alternative here ?

**Reply by jsjoeio** [reply]
Hmm... one alternative is to tell your users they must use Chrome. I don't like that personally. You probably need to re-write your code to support all browsers, which means you might need to use the `navigator` API directly instead of the VS Code one. That's the route I would take personally. I dug into the VS Code types and it doesn't seem like their API has something like `requestClipboardPermissions`.

> For example, Firefox does not yet support the "clipboard-read" and "clipboard-write" permissions, so access to the methods that access and change the contents of the clipboard are restricted in other ways.

Hmm...seems like a blocker. 

> Note: At this time, while Firefox does implement read(), it does not recognize the "clipboard-read" permission, so attempting to use the [Permissions API](https://developer.mozilla.org/en-US/docs/Web/API/Permissions_API) to manage access to the API will not work.

Yeah hmm...I think you might need to use a different solution then. This seems like a blocker at the browser level.

[source](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard/read)

**Reply by Oliver-Weng-Sorcerer** [reply]
@jsjoeio , yup, that makes sense, thank you for the brainstorming and info sharing. helps a lot.

**Reply by jsjoeio** [reply]
This was actually really fun because it was specific and something I've worked with before so thanks for letting me brainstorm with you! Good luck!
